### ISSUE TITLE
Add a vjs.extend function to videoJs

### ISSUE THREAD

**Issue by carpasse** [videojs/video.js#1856]
Hi guys 

I really miss an extend function as part of the core of videoJS. I have seen many videoJs plugins with a really simple extend function at the beginning of it.

I am willing to do the PR myself if you guys think it is worth it.

What do you think?

**Comment by heff** [comment]
Are you wanting something like our [object merge](https://github.com/videojs/video.js/blob/master/src/js/lib.js#L98) (not exported)? 

Adding additional APIs will be part of the 5.0 tasks, and I imagine this will be included.

**Comment by carpasse** [comment]
Unfortunately object merge is not good enough in my opinion. 

First it doesn't allow for multiple argument maps.
I would like to do something like: 

 ```
var options = vjs.extend({}, defaultOpt, customOptions);
```

also to use vjs.obj.merge to generate options on a case like the one above is dangerous and can lead to unexpected bugs.

for example in a situation like the one below:

```
var defaultOptions = {
    profession: {
       name: 'developer',
       level: 'senior'
    },
    language: {
         english: 'native'
         spanish: 'good'
    }
};

var options = vjs.obj.merge({}, defaultOptions);

// If I now change the options I may end up changing the default options also
options.profession.name = 'Project Manager';

//I would expect the variable below to be false but with vjs.obj.merge returns true
var isSameProfession = defaultOptions.profession.name === options.proffession.name;
```
I also believe that merge is just too shallow, if you do:

```
var languages = {    
    language: {
         english: 'native'
         spanish: 'good'
    }
};

vjs.merge(languages, { language: {french: 'medium'} } );
```
I would expect languages to end up like: 
```
 {    
    language: {
         english: 'native'
         spanish: 'good',
         french: 'medium'
    }
}
```

but what we get is:

```
 {    
    language: {
         french: 'medium'
    }
}
```

vjs.obj.merge is not good unless you are merging 2 very simple key value maps where the values are just primitive values like string or number.

Therefore my request to add a proper 'extend' method to the core of videoJS.

**Comment by heff** [comment]
@carpasse I think this makes sense, and I'd be interested to see your implementation of it if you're still interested in making a pull request for it.

@videojs/core-committers any other opinions on this one?

**Comment by gkatsev** [comment]
We also have a `deepMerge` already in lib.
If we add extend, it should just be a wrapper for `merge` and `deepMerge` with variable arity.

**Comment by carpasse** [comment]
Wow. I missed this deepMerge method. 
I will try to do the pull request using deepMerge today or tomorrow.

Thanks.

**Comment by mmcc** [comment]
:thumbsup: sounds good, Carlos! Thanks for the help. Please reference this issue with the PR you open so we can remember to close it (or use the `closes #1856` syntax in a commit message).

**Comment by carpasse** [comment]
@mmcc I forgot to reference the pull request with this issue. I tried to change the name of the pull request adding fixes #1856 but it didn't work.

Sorry, you can find the pull here https://github.com/videojs/video.js/pull/1909

@gkatsev I tried to use deepMerge. 

Since I couldn't find any test for it. I wrote some to ensure that no reference to the original object was being preserved on the merged obj.

Unfortunately that was not the case. I changed it but when I did. I got a lot of tests failing so I decided to leave it as it is.

**Comment by mmcc** [comment]
Pretty sure this can be closed due to other PRs that got merged, but correct me if I'm wrong and we can re-open.
